### PR TITLE
Add numeric_stable_mode parameters to softmax_with_xe op

### PR DESCRIPTION
**Pull request by sneaxiy**
When numeric_stable_mode = True, the implementation is more stable in numeric. 
When numeric_stable_mode = False, the implementation is based on cudnn in GPU.

Curiously, the speed of numeric_stable_mode = True is faster when dims are small.

Profiler Results:

-----------------------------------------------------------------------------
**batch_size = 200, feature_size = 400**

When numeric_stable_mode = False
```
------------------------->     Profiling Report     <-------------------------

Place: All
Time unit: ms
Sorted by total time in descending order in the same thread

Event                                  Calls       Total       Min.        Max.        Ave.        Ratio.      
thread0::softmax_with_cross_entropy    100         11.1597     0.102848    0.20112     0.111597    0.833096    
```

When numeric_stable_mode = True
```
------------------------->     Profiling Report     <-------------------------

Place: All
Time unit: ms
Sorted by total time in descending order in the same thread

Event                                  Calls       Total       Min.        Max.        Ave.        Ratio.      
thread0::softmax_with_cross_entropy    100         10.1696     0.089824    0.152896    0.101696    0.830787    
```

-----------------------------------------------------------------------------
**batch_size = 1000, feature_size = 1000**

When numeric_stable_mode = False
```
------------------------->     Profiling Report     <-------------------------

Place: All
Time unit: ms
Sorted by total time in descending order in the same thread

Event                                  Calls       Total       Min.        Max.        Ave.        Ratio.      
thread0::softmax_with_cross_entropy    100         13.2217     0.12064     0.170208    0.132217    0.85466     
 ```

When numeric_stable_mode = True

```
------------------------->     Profiling Report     <-------------------------

Place: All
Time unit: ms
Sorted by total time in descending order in the same thread

Event                                  Calls       Total       Min.        Max.        Ave.        Ratio.      
thread0::softmax_with_cross_entropy    100         18.7698     0.166624    0.23232     0.187698    0.88331     
```

### CODE REVIEW

**Review by reyoung — DISMISSED** [review]

File: paddle/fluid/operators/softmax_with_cross_entropy_op.cu [hunk 1]
```diff
@@ -194,11 +205,136 @@ __global__ void RowReductionForSoftmaxAndCrossEntropy(const T* logits_data,
...
+        logits_data, loss_data, softmax_data, feature_size);                 \
+    if (ignore_idx >= 0 && ignore_idx < feature_size) {                      \
+      platform::ForRange<platform::CUDADeviceContext> for_range(             \
+          ctx, batch_size* feature_size);                                    \
+      for_range(HardLabelSoftmaxWithCrossEntropyFunctorWithIgnoreIdx<T>(     \
+          logits_data, labels_data, loss_data, softmax_data, feature_size,   \
+          ignore_idx));                                                      \
+    } else {                                                                 \
+      platform::ForRange<platform::CUDADeviceContext> for_range(             \
+          ctx, batch_size* feature_size);                                    \
```
Comment (chengduoZH) [inline paddle/fluid/operators/softmax_with_cross_entropy_op.cu]: line 307~308 can be placed out of `if ... else ...`.

**Review by qingqing01 — DISMISSED** [review]

File: paddle/fluid/API.spec [hunk 2]
```diff
@@ -103,7 +103,7 @@ paddle.fluid.layers.beam_search ArgSpec(args=['pre_ids', 'pre_scores', 'ids', 's
 paddle.fluid.layers.row_conv ArgSpec(args=['input', 'future_context_size', 'param_attr', 'act'], varargs=None, keywords=None, defaults=(None, None))
 paddle.fluid.layers.multiplex ArgSpec(args=['inputs', 'index'], varargs=None, keywords=None, defaults=None)
 paddle.fluid.layers.layer_norm ArgSpec(args=['input', 'scale', 'shift', 'begin_norm_axis', 'epsilon', 'param_attr', 'bias_attr', 'act', 'name'], varargs=None, keywords=None, defaults=(True, True, 1, 1e-05, None, None, None, None))
-paddle.fluid.layers.softmax_with_cross_entropy ArgSpec(args=['logits', 'label', 'soft_label', 'ignore_index'], varargs=None, keywords=None, defaults=(False, -100))
+paddle.fluid.layers.softmax_with_cross_entropy ArgSpec(args=['logits', 'label', 'soft_label', 'ignore_index', 'use_cudnn'], varargs=None, keywords=None, defaults=(False, -100, True))
```
Comment (panyx0718) [inline paddle/fluid/API.spec]: I don't like putting `use_cudnn` in the API. what if there is other nvidia library, or mkl, mkldnn, ngraph in the future? are we going to put more use_xxx in the api?

**Review by panyx0718 — CHANGES_REQUESTED** [review]
Can you just make `use_cudnn` the default one? and don't update api? I don't think user will know use_cudnn=False is more stable and use it.

**Review by panyx0718 — APPROVED** [review]
LG API

**Review by qingqing01 — APPROVED** [review]

**Pull request merged**